airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/message/DestinationMessage.kt

```diff
@@ -200,7 +205,62 @@ data class DestinationRecordRaw(
     }
 
     fun asEnrichedDestinationRecordAirbyteValue(): EnrichedDestinationRecordAirbyteValue {
-        TODO()
+        val rawJson = asRawJson()
+
+        // Get the set of field names defined in the schema
+        val schemaFields =
+            when (schema) {
+                is ObjectType -> schema.properties.keys
+                else -> emptySet()
+            }
+
+        val declaredFields = mutableMapOf<String, EnrichedAirbyteValue>()
+        val airbyteMetaFields = mutableMapOf<String, EnrichedAirbyteValue>()
+        val undeclaredFields = mutableMapOf<String, JsonNode>()
+
+        // Process fields from the raw JSON
+        rawJson.fields().forEach { (fieldName, fieldValue) ->
+            when {
+                schemaFields.contains(fieldName) -> {
+                    // Declared field (exists in schema)
+                    val fieldType =
+                        (schema as ObjectType).properties[fieldName]?.type
+                            ?: throw IllegalStateException(
+                                "Field '$fieldName' exists in schema keys but not in properties"
+                            )
+
+                    val airbyteValue =
+                        if (fieldValue.isNull) NullValue else fieldValue.toAirbyteValue()
+                    declaredFields[fieldName] =
+                        EnrichedAirbyteValue(
+                            value = airbyteValue,
+                            type = fieldType,
+                            name = fieldName,
+                            fieldCategory = FieldCategory.CLIENT_DATA
+                        )
+                }
+                else -> {
+                    // Undeclared field (not in schema)
+                    undeclaredFields[fieldName] = fieldValue
+                }
+            }
+        }
+
+        return EnrichedDestinationRecordAirbyteValue(
+            stream = stream,
+            declaredFields = declaredFields,
+            airbyteMetaFields = airbyteMetaFields,
```

Comment:
starting to wonder if airbyteMetaFIelds shouldn't even be an arg (maybe this is what you were referring to by extension method)
```kotlin
class EnrichedDestinationRecordAirbyteValue(...) {
  val airbyteMetaFields = mapOf(
    "airbyte_meta" to EnrichedAirbyteValue(...),
    "airbyte_extracted_at" to EnrichedAirbyteValue(...)
  )
}
```